### ISSUE TITLE
Report within BBMap tools request: pileup.sh

### ISSUE THREAD

**Issue by mgruca13** [ewels/MultiQC#768]
* **Name of tool:**
     BBMap : pilup.sh
* **Tool description:**
    This tool is a useful quick run of bp coverage/GC content/positive and negative strand coverage, etc.
* **Tool homepage:**
    https://github.com/BioInfoTools/BBMap/blob/master/sh/pileup.sh
File log output:

[41_DMSO_1.coverage.stats.txt](https://github.com/ewels/MultiQC/files/2040755/41_DMSO_1.coverage.stats.txt)

* **Log filename pattern:**
    * [eg. `*_mytool.txt`]
    
* **Most interesting data for General Stats table:**
    * percentage coverage by chromosome, GC content
* **Data suitable for MultiQC plot(s):**
    * Histogram overlay of different features

Thanks for your consideration and efforts!

**Comment by ewels** [comment]
Hi @mgruca13,

Thanks for this! You've said that the filename pattern is `coverage.txt`, in that this tool always produces files called `coverage.txt`. But the example file you attached is called `41_DMSO_1.coverage.stats.txt`.

Could you clarify what the output files from this tool are called please? I guess that there is no filename pattern - from the top of my head I think that's what BBMap usually does.

@boulund - here's another bbmap submodule for you if you have time 😉 

Phil

**Comment by boulund** [comment]
Great fun! 

I was quite hesitant to implement support for both `pileup.sh` and `bbsplit.sh` because they are very general and produce for each reference sequence, and the total volume can vary immensely. Imagine two reference genomes -- no biggie, then all chromosomes of the human genome -- also not really an issue, but then someone (like me) comes along with reference gene catalogs of 10M sequences -- now the module needs to handle a wide variety of output visualizations for all these common scenarios. I just didn't have the time nor energy to come up with something general enough to handle these different sizes of output.

However, the work I'm planning to put into visualizing BBSplit output should be very easy to transfer to Pileup as well. Let's hope for the best :D 🥇

**Comment by ewels** [comment]
I guess you could do a stacked barplot of those two statistics, one bar per sample, one category per chromosome. It'll be a lot of categories so pretty ugly, but should highlight any samples that don't follow the pattern.

It's probably worth allowing module config patterns to ignore chromosomes by name pattern and length (eg. skipping anything under 1 Mbp by default should keep things fairly clean).

Could even be clever and sum the total `Length` and only show chromosomes that make up at least `x%` (5%?) of the total, binning the rest into a single category.

**Comment by ewels** [comment]
Also, instead of a stacked barplot you could do a line plot with categories on the X axis for each chromosome. This would probably be better actually.

**Comment by boulund** [comment]
Sounds like a fun idea, I'll definitely play around with that.

**Comment by ewels** [comment]
Hi @mgruca13,

Thanks for the context - that's helpful! If you could attach the histogram data as a file through the GitHub web interface that would be great. I think it has been truncated here, plus GitHub can change the whitespace sometimes, so file uploads are always more reliable.

Cheers,

Phil

**Comment by ewels** [comment]
@boulund - I found the module that does this coverage-per-chromosome plot already. It's samtools idxstats. I imagine you can use much of the code that's already there..

**Comment by boulund** [comment]
Super, thanks for doing that research! Does that already handle situations with up to tens of thousands of "chromosomes"?

**Comment by ewels** [comment]
Yup - it goes through the chromosomes one by one until it has represented 99.9% of reads (default, user configurable). It also has user-config keys for chromosomes to always show and always hide chromosomes, plus to define which chromosomes are X / Y for a sex prediction plot.

It's not the cleanest code ever, but it's functional 😉 

https://github.com/ewels/MultiQC/blob/2d873203de31dbfe3ff0bfc3852c040937a4b212/multiqc/modules/samtools/idxstats.py#L36-L104

https://github.com/ewels/MultiQC/blob/2d873203de31dbfe3ff0bfc3852c040937a4b212/multiqc/modules/samtools/idxstats.py#L126-L145
